src/clib/pio_msg.c

```diff
@@ -1616,61 +1645,602 @@ int pio_msg_handler(int io_rank, int component_count, iosystem_desc_t *iosys)
+
+    /* Find the rank of the io leader in peer_comm. */
+    if (!my_iosys->io_rank)
+    {
+	if ((mpierr = MPI_Comm_rank(peer_comm, &iam)))
+	    ierr = check_mpi(NULL, mpierr, __FILE__, __LINE__);
+    }
+    else
+	iam = -1;
+
+    /* Find the lucky io_leader task. This matches a call in
+     * init_comp_comm(). */
+    LOG((3, "calling Allreduce for io_leader = %d", *io_leader));
+    if ((mpierr = MPI_Allreduce(&iam, io_leader, 1, MPI_INT, MPI_MAX, peer_comm)))
+	ierr = check_mpi(NULL, mpierr, __FILE__, __LINE__);
+    LOG((3, "called Allreduce for io_leader = %d", *io_leader));
+
+    /* Find the rank of the comp leader in peer_comm. */
+    iam = -1;
+    if ((mpierr = MPI_Allreduce(&iam, comp_leader, 1, MPI_INT, MPI_MAX, peer_comm)))
+	ierr = check_mpi(NULL, mpierr, __FILE__, __LINE__);
+
+    /* This is an io task. */
+    my_iosys->ioproc = true;
+
+    /* Is this the iomaster? Only if the io_rank is zero. */
+    my_iosys->iomaster = !my_iosys->io_rank ? MPI_ROOT : MPI_PROC_NULL;
+    LOG((3, "I am IO %s!", my_iosys->iomaster == MPI_ROOT ? "MASTER" : "SERVANT"));
+
+    /* Set up the intercomm from the I/O side. */
+    LOG((3, "about to call MPI_Intercomm_create my_iosys->io_comm = %d comp_leader = %d cmp = %d",
+	 my_iosys->io_comm, *comp_leader, cmp));
+    if ((mpierr = MPI_Intercomm_create(my_iosys->io_comm, 0, peer_comm,
+				       *comp_leader, cmp, &my_iosys->intercomm)))
+	ierr = check_mpi(NULL, mpierr, __FILE__, __LINE__);
+    LOG((3, "intercomm created!"));
+
+    /* Create the union communicator. */
+    if ((mpierr = MPI_Intercomm_merge(my_iosys->intercomm, 0, &my_iosys->union_comm)))
+	ierr = check_mpi(NULL, mpierr, __FILE__, __LINE__);
+    LOG((3, "union communicator created"));
+
+    return ierr;
+}
+
+/**
+ * Initialize a computation communicator. This is only run on tasks
+ * that are part of a computation component. It is run for each
+ * computation component, on each computational task. When more than
+ * one computational coponent is involved, the comp_comms[] element
+ * for that component will be NULL except for the compent on which it
+ * applies.
+ *
+ */
 int
-pio_iosys_print(int my_rank, iosystem_desc_t *iosys)
+init_comp_comm(MPI_Comm comp_comm, iosystem_desc_t *my_iosys, MPI_Comm peer_comm,
+	       int cmp, int *comp_leader, int *io_leader)
 {
-    printf("%d iosysid: %d\n", my_rank, iosys->iosysid);
-    if (iosys->union_comm == MPI_COMM_NULL)
-	printf("%d union_comm: MPI_COMM_NULL ", my_rank);
+    MPI_Group peer_group;
+    int iam;
+    int mpierr;
+    int ierr = PIO_NOERR;
+
+    LOG((2, "init_comp_comm comp_comm = %d peer_comm = %d cmp = %d",
+	 comp_comm, peer_comm, cmp));
+
+    /* Set these values for all comp comms. */
+    my_iosys->io_comm = MPI_COMM_NULL;
+    my_iosys->iogroup = MPI_GROUP_NULL;
+    my_iosys->io_rank = -1;
+    my_iosys->ioproc = false;
+    my_iosys->iomaster = MPI_PROC_NULL;
+
+    /* If multiple components are in use, this function will be called
```

Comment:
Stand by while I work on this a bit.

Can you explain the use case for having a processor involved in more than one component? Why would you want to do that?

Thanks,
Ed